### PR TITLE
Reworked Items

### PR DESCRIPTION
**Pull request by d2inventory**
All the items of the player are now in Items.
The Location attribute of Item is not just BodyLocation anymore,
it contains the position in the inventory along with it's width and height.
The Container indicates where the Item is stored (Equipment, Belt, Cube, Inventory, Stash)

Changed GetItemDescription to take a D2Unit (item) instead of Item and made it public
Added Location for hireling items

### CODE REVIEW

**Comment by d2inventory** [comment]
It works on my end, but still needs to be tested with Plugy and PD2 etc.

Also maybe the Hireling needs to be done differently. Also nothing done for the PipeServer thingy :D

File: src/D2Reader/GameMemoryTableFactory.cs [hunk 1]
```diff
@@ -68,30 +68,30 @@ private GameMemoryTable CreateMemoryTableForVersion114D(IntPtr BaseAddress)
...
+                Units113                    = null,
+                Units114                    = BaseAddress + 0x3A5E70,
+                PlayerUnit                  = BaseAddress + 0x3A5E74,
+                Area                        = BaseAddress + 0x3A3140,
+                Pets                        = BaseAddress + 0x3BB5BC,
+                InventoryTab                = BaseAddress + 0x3BCC4C,
+                StringIndexerTable          = BaseAddress + 0x4829B4,
+                StringAddressTable          = BaseAddress + 0x4829B8,
+                PatchStringIndexerTable     = BaseAddress + 0x4829D0,
+                PatchStringAddressTable     = BaseAddress + 0x4829BC,
```
Comment (Zutatensuppe) [inline src/D2Reader/GameMemoryTableFactory.cs:94]: there is no change here apart from the formatting
the file should not be part of the PR

maybe we can implement some kind of coding standard and then reformat all the code accordingly, but for now i'd like to keep this as it is, if there is no real code change

File: src/D2Reader/Models/Item.cs [hunk 2]
```diff
@@ -18,8 +19,50 @@ public class Item
         internal bool IsEthereal() => HasFlag(ItemFlag.Ethereal);
         internal bool HasRuneWord() => HasFlag(ItemFlag.Runeword);
 
-        internal bool IsInCube() => ItemData.InvPage == InventoryPage.HoradricCube;
-        internal bool IsInStash() => ItemData.InvPage == InventoryPage.Stash;
-        internal bool IsInInventory() => ItemData.InvPage == InventoryPage.Inventory;
+        internal bool IsInCube() => !IsEquipped() && ItemData.InvPage == InventoryPage.HoradricCube;
+        internal bool IsInStash() => !IsEquipped() && ItemData.InvPage == InventoryPage.Stash;
+        internal bool IsInInventory() => !IsEquipped() && ItemData.InvPage == InventoryPage.Inventory;
```
Comment (Zutatensuppe) [inline src/D2Reader/Models/Item.cs:24]: Do you know why IsEquipped check is needed in all those cases?

File: src/D2Reader/Struct/Inventory/D2InventoryInfo.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
+using System.Runtime.InteropServices;
```
Comment (Zutatensuppe) [inline src/D2Reader/Struct/Inventory/D2InventoryInfo.cs:1]: nice job finding the new structures =D

File: src/D2Reader/Readers/UnitReader.cs [hunk 4]
```diff
@@ -139,9 +139,9 @@ public D2MonsterData GetMonsterData(D2Unit unit)
             return reader.Read<D2MonsterData>(unit.UnitData);
         }
 
-        private D2ItemDescription GetItemDescription(Item item)
+        public D2ItemDescription GetItemDescription(D2Unit item)
```
Comment (Zutatensuppe) [inline src/D2Reader/Readers/UnitReader.cs:142]: Fine for now =)

But I d prefer if this could stay Item (more clear that 100% item comes in, and never another unit), but the change in InventoryReader makes this necessary atm.

File: src/D2Reader/Models/Item.cs [hunk 5]
```diff
@@ -18,8 +19,50 @@ public class Item
...
+            if (obj == null || this.GetType() != obj.GetType())
+                return false;
+
+            var other = (ItemLocation)obj;
+            return this.X            == other.X &&
+                   this.Y            == other.Y &&
+                   this.Width        == other.Width &&
+                   this.Height       == other.Height &&
+                   this.BodyLocation == other.BodyLocation &&
+                   this.Container    == other.Container;
```
Comment (Zutatensuppe) [inline src/D2Reader/Models/Item.cs:60]: No need to change, but just a note because i saw it in multiple places:

So far we didn't use this kind of alignment of parameters or operators, and at least the reason for me is that a change in one line may increase PR unecessary, because all lines need to be adjusted if longest line changes. But as said, no need to change, it may be better readable this way, and we read this more often than change it :D

And as this is new code and not purely formatting, this is fine for me!

File: src/D2Reader/Models/Item.cs [hunk 6]
```diff
@@ -18,8 +19,50 @@ public class Item
...
+        }
+
+        public override int GetHashCode()
+        {
+            return X.GetHashCode() ^ Y.GetHashCode() ^ Width.GetHashCode() ^ Height.GetHashCode() ^ BodyLocation.GetHashCode() ^ Container.GetHashCode();
```
Comment (Zutatensuppe) [inline src/D2Reader/Models/Item.cs:65]: pretty long line ^^

**Review by Zutatensuppe — COMMENTED** [review]
please see comments =)
I would ask to undo the 'formatting only' change in GameMemoryTableFactory for now. (Probably the whole project code should be reformatted as a whole, if something like this is done, so that it is consistent. I think it is not 100% consistent atm anyway, but doing partly reformat will make it more inconsistent ^^)

All the other comments are just questions and comments that don't require a change now in my opinion.

**Comment by d2inventory** [comment]
Imma reply to all in this comment.

I would prefer it if the GameMemoryTableFactory.cs would be all formated like that because it very quickly highlights the memory region the data is in. I formatted that because I initially added my own pointers to get to the item containers and frankly it was horrible to read that code with some addresses normalized to 8 digits and everything "jagged".
This is comming from a perspective of me being a complete stranger to the code and wanting to quickly read into it to get the information I am looking for.
To me this type of information is "spread sheet data" so imo it should be formatted like a spread sheet.
I got it all formatted like this on my end, so I can do either one, remove the file all together from the PR or update it with consistent formatting like the example shown.

I am not sure why I added the IsEquipped() check, probably just didn't understand the way it was done at the time. It doesn't really make any sense if I am looking at it now so I guess I just remove that part.
The functions aren't in use anymore anyway but I thought I leave them in there cuz why not 😄 

GetItemDescription()
I kinda disagree here, because the class is UnitReader and imo it should perform actions around units so it handling a D2Unit of type item makes more sense than an Item directly where the Unit has to be extracted first.
I'm more "annoyed" because I had to make it public. I think the UnitReader should offer a single point solution to get the DiabloInterface representation of an item from a D2Unit (be it a Player to get multiple items or one D2Unit-Item to Item "converter").
But yeah, maybe that's just me misunderstanding what the goal is.
I fully agree though that there needs to be a check that no other D2Unit than one of type Item can be passed to the functions. So maybe add something like this to the top of the function?
´´´
if (item.eType != D2UnitType.Item)
    throw new Exception("Wrong D2Unit type");
´´´

Item Comparison
I already said why I like to format "spread sheet data" in this style.
The line is indeed very long hahaha, one ugly motherfucker 😄 
Tbh this is just boilerplate code I copy pasted to satisfy the compiler.
Waddayathink should I just break it into 2 lines to fit more with all the other code?

**Comment by d2inventory** [comment]
alrighty, I reverted the formatting in GameMemoryTableFactory and removed the unnecessary checks in the Item class

I also added the check in GetItemDescription so there will be a sane Exception if the wrong D2Unit type is passed ot the function.

**Review by Zutatensuppe — APPROVED** [review]
:tada: nice work! =)
